torch/csrc/jit/codegen/cuda/transform_view.cpp

```diff
@@ -0,0 +1,678 @@
+          original_view_[state_.original_view_index];
+
+      if (kCurrentSize == kEmptyAxis) {
+        // If current size in transform view is 0, then it was already handled
+        // and should be skipped.
+        ++state_.transform_view_index;
+      } else if (kCurrentSize == new_view_[state_.new_view_index]) {
+        addKeepTransform(kRFactor);
+      } else if (new_view_[state_.new_view_index] == kSingletonAxis) {
+        addBroadcastTransform();
+      } else {
+        MergeThenSplit::process(
+            this,
+            state_,
+            kCurrentSize,
+            new_view_[state_.new_view_index],
+            kRFactor);
+      }
+    }
+
+    MergeAdjacentSingletonDimensions::process(this, state_);
+
+    // Skip any root domains that were merged for any splits with remainder
+    // OR any singleton axes
+    while (state_.transform_view_index < transform_view_.size() &&
+           transform_view_[state_.transform_view_index] <= kSingletonAxis) {
+      ++state_.transform_view_index;
+    }
+
+    // Add broadcast axes for any remaining size 1 dimensions
+    while (state_.original_view_index == original_view_.size() &&
+           state_.new_view_index < new_view_.size() &&
+           new_view_[state_.new_view_index] == kSingletonAxis) {
+      addBroadcastTransform();
+    }
+  }
+
+ private:
+  ViewIndexState state_;
+
+  std::vector<ViewTransform*> view_transforms_;
+  std::vector<BroadcastTransform*> broadcast_transforms_;
+  std::vector<TrivialReductionTransform*> trivial_reduction_transforms_;
+
+  const std::vector<IterDomain*> root_domain_;
+  const std::vector<size_t>& original_view_;
+  const std::vector<size_t>& new_view_;
+
+  // transform_view is a mutable view and is initialized with the original_view.
+  // It is used to track the current state of the original tensor domain.
+  //
+  // When we merge dimensions in the original_view, we multiply the sizes of
+  // the adjacent, merged dimensions. The product size is placed in the current
+  // position of the transform_view, while the other dimensions are set to 0.
+  //
+  // When we add a Split transform, the current size is divided by the outer
+  // split factor.
+  //
+  // Size-0 dimensions are automatically skipped.
+  //
+  // If transform size != original size for an axis, then the transformation
+  // uses the last rfactor domain. Otherwise, it is a root domain
+  // transformation.
+  std::vector<size_t> transform_view_;
+};
+
+//! Create new TensorDomain with a modified rfactor domain using the specified
+//! view transformations
+TensorDomain* createViewDomain(
+    TensorDomain* original_domain,
+    const std::vector<ViewTransform*>& view_transforms) {
+  FUSER_PERF_SCOPE("createViewDomain");
+
+  TORCH_INTERNAL_ASSERT(!view_transforms.empty());
+
+  std::vector<IterDomain*> new_root_domain;
+  for (auto id : TensorDomain::noReductions(original_domain->getRootDomain())) {
```

Comment:
```suggestion
  for (auto id : TensorDomain::noReductions(original_domain->getMaybeRFactorDomain()))
```
I understand view of view isn't supported at this point, but this would just work when it becomes supported.